cxx-qt-gen/src/cxx_qt.h

```diff
@@ -73,53 +84,100 @@ class CxxQObject : public QObject
+    }
+
+    m_queue.push_back(std::make_pair(QueueEvent::EmitSignal, signalFunctor));
+  }
+
+  void requestPropertyChange(std::function<void()> propertyFunctor)
+  {
+    // Lock the queue, post the event, add to the queue
+    // worst case we'll push an event that does nothing if takeQueue() is
+    // waiting on the lock
+    const std::lock_guard<std::mutex> guard(m_queueMutex);
+
+    if (!m_waitingForUpdate.exchange(true, std::memory_order_relaxed)) {
+      QCoreApplication::postEvent(this, new QEvent(ProcessQueueEvent));
+    }
+
+    m_queue.push_back(
+      std::make_pair(QueueEvent::UpdatePropertyChange, propertyFunctor));
+  }
+
+  void requestUpdate(std::function<void()> updateFunctor)
+  {
+    // Lock the queue, post the event, add to the queue
+    // worst case we'll push an event that does nothing if takeQueue() is
+    // waiting on the lock
+    const std::lock_guard<std::mutex> guard(m_queueMutex);
+
+    if (!m_waitingForUpdate.exchange(true, std::memory_order_relaxed)) {
+      QCoreApplication::postEvent(this, new QEvent(ProcessQueueEvent));
+    }
+
+    // Compress request updates to only one
+    //
+    // TODO: should we compress events? what happens if we change a
+    // property/emit signal, request update, change property/emit signal,
+    // request update? The second request update won't happen?
+    // Or should the request update always happen after the property/signal
+    // changes?
+    if (std::none_of(
+          m_queue.cbegin(),
+          m_queue.cend(),
+          [](const std::pair<QueueEvent, std::function<void()>>& item) {
+            return item.first == QueueEvent::UpdateState;
+          })) {
+      m_queue.push_back(std::make_pair(QueueEvent::UpdateState, updateFunctor));
+    }
+  }
+
+  std::vector<std::pair<QueueEvent, std::function<void()>>> takeQueue()
+  {
+    const std::lock_guard<std::mutex> guard(m_queueMutex);
+    std::vector<std::pair<QueueEvent, std::function<void()>>> queue;
+    std::swap(m_queue, queue);
+    return queue;
   }
 
   bool event(QEvent* event) override
   {
-    if (event->type() == UpdateStateEvent) {
-      // New Rust-side events might come in while we are processing a request
-      // and request a new update while we are processing the queue.
+    if (event->type() == ProcessQueueEvent) {
+      // New Rust-side events might come in while we are processing the queue.
       //
-      // If we flip this flag before updateState then worst case we get an extra
-      // event with nothing to actually process whereas if we do it afterwards
-      // then we might miss an update request.
+      // If we flip this flag before takeQueue then worst case we get an
+      // extra event with nothing to actually process whereas if we do it
+      // afterwards then we might miss a queue item to process.
       m_waitingForUpdate.store(false, std::memory_order_relaxed);
 
-      updateState();
+      for (auto item : takeQueue()) {
```

Comment:
Maybe use a structured binding here, i.e. `const auto& [type, functor] : takeQueue()`. Makes it more readable.

Reply:
Now we don't have a pair, but do you want this to change? should it be a `const auto&` ?

```
    for (auto item : takeQueue()) {
      item();
    }
```

Reply:
Not sure if the compiler can optimize this away, but simply using `auto` would likely cause the compiler to copy every std::function. So using `const auto& item` would indeed be preferred here.

Reply:
Changed :-)